directives part-two

jgraettinger opened the pull request:
This is the second major chunk of directives work, and includes full user onboarding including stats materializations.

Please review commit by commit.

Next:
 * Testing with pgBouncer to confirm we can use it to host _lots_ of these materializations. Probably tweaking the materialize-postures connector along the way (no impact on this PR in particular).
 * Working up and testing a migration from the production database.
 * Updating production release process to point at ops tenant template by-URL.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/estuary/animated-carnival/68)
<!-- Reviewable:end -->

psFried commented on supabase/migrations/03_catalog-types.sql:
```diff
@@ -24,6 +24,20 @@ For example: "acmeCo/anvils/" or "acmeCo/products/TnT_v4/",
 
+create domain catalog_tenant as text
+  constraint "Must be a valid catalog tenant"
+  check (value ~ '^[[:alpha:][:digit:]\-_.]+/$' and value is nfkc normalized);
```
Doesn't have to be right now, but I think we should consider additional restrictions on tenant names. As it is, I don't see how we'd disallow names like `../` or `-/`, which seem likely to cause issues. I'm also still skeptical of allowing both `foo/` and `Foo/`, since it could easily lead to confusion for users.

jgraettinger replied:
The `tenant_exists` check does case-invariant comparison, so you can't have `foo/` and `Foo/`. Once provisioned, the tenant is case sensitive though.

We can also "reserve" tenants by creating them in the `tenants` table without being hooked up to anything.

psFried reviewed (commented):
Just commenting now, since there's still some work TBD

jshearer commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
+}
+
+pub async fn provision_tenant(
+    accounts_user_email: &str,
```
Why email here, but `tenant_user_id` down below?

jgraettinger replied:
accounts_user_email is, for example, `accounts@estuary.dev`. It's a separate account from the user ID on whose behalf we're acting.

jshearer commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
-        r#"with s1 as (
+    let provisioned = sqlx::query_as!(
+        ProvisionedTenant,
+        r#"with
```
Are all of these CTEs the same as/equivalent to using a transaction? My concern would be leaving things in a broken state if something fails here, and we don't automatically rollback

jgraettinger replied:
We're already in a transaction within this context, so they could safely be separate statements.

The CTE's are for tighter narrative structure / less boilerplate. It's slightly more efficient to boot but that doesn't matter here.

jshearer commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
+    // Create partition of task_stats which will home all stats of the tenant.
+    sqlx::query(&format!(
+        r#"
+        create table task_stat_partitions."{tenant}_stats"
```
It wasn't as much of a concern above because of bound input parameters, but since we're directly templating variable into SQL here, do we have to worry about injection? Can I ask for a `tenant` of `hello"; drop table users; --`, or some such?

jgraettinger replied:
No, at this stage `tenant` has had a lot of validation. It's unicode letters, decimals, `_`, `-`, `.` and that's it.

jshearer commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
+
+    // Create partition of task_stats which will home all stats of the tenant.
+    sqlx::query(&format!(
+        r#"
```
Woooah cool, Rust has inline format strings now!

jshearer commented on crates/agent/src/directives/mod.rs:
```diff
@@ -40,11 +40,18 @@ pub enum Directive {
+#[derive(Default)]
+pub struct DirectiveHandler {
+    tenant_template: models::Catalog,
+    accounts_user_email: String,
```
I see this passed around a bunch... this is `accounts@estuary.dev` right? Should this just be a global value anyone can reference? Maybe based on env vars or something? It doesn't seem like it'll ever change during the lifecycle of any of these crates

jgraettinger replied:
Yep, `accounts@estuary.dev` or `accounts@company-who-self-hosts.com`, etc.

jshearer commented on crates/agent/src/main.rs:
```diff
@@ -37,10 +37,22 @@ struct Args {
     #[clap(long = "bin-dir", env = "BIN_DIR")]
     bindir: String,
+    /// Email address of user which provisions and maintains tenant accounts.
+    #[clap(long = "accounts-email", default_value = "accounts@estuary.dev")]
```
Ah, I see. I suppose ignore my above comment about `accounts_email`... though it is too bad that it has to be threaded through quite so many calls. `¯\_(ツ)_/¯`

jgraettinger commented on supabase/migrations/11_stats.sql:
```diff
@@ -1,85 +1,55 @@
-    'stats for each task shard aggregated by the day';
-grant select on task_stats_by_day to authenticated;
-
+create table task_stats (
```
Re: meta tables, that's correct, they're created in the `task_stat_partitions` schema and are hidden from the public API.

I'm working up and testing a migration.

jgraettinger commented on crates/flowctl/src/raw.rs:
```diff
@@ -42,6 +44,20 @@ pub struct Get {
+    #[clap(long)]
+    table: String,
+    /// Optional query parameters.
+    #[clap(short = 'q', parse(try_from_str = parse_key_val), number_of_values = 1)]
```
Done.

jgraettinger commented on ops-catalog/template-common.flow.yaml:
```diff
@@ -0,0 +1,112 @@
+        documents:
+          - shard: &mappedShard
+              kind: capture
+              name: tenant/test/cap
```
Done.

jgraettinger commented on ops-catalog/template-local.flow.yaml:
```diff
@@ -0,0 +1,22 @@
+    bindings:
+      - source: ops/TENANT/stats-by-hour
+        resource:
+          table: TENANT_stats
```
Yep, it will.

jgraettinger commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
+            -- which will then fail a not-null constraint.
+            select (select id from auth.users where email = $4 limit 1) as accounts_id
+        ),
+        grant_user_admin_to_tenant as (
```
Yep, CTEs are not lazily evaluated.

It could be separate statements. Went with this for narrative cohesion and reduced verbosity.

jgraettinger commented:
PTAL. I've addressed feedback here and have introduced a full, tested migration script for moving from a checked-in stack (or production) to a post-directives install. 🎉

psFried commented on crates/agent-sql/src/directives/beta_onboard.rs:
```diff
@@ -41,39 +41,89 @@ pub async fn tenant_exists(
+        ),
+        create_storage_mappings as (
+            insert into storage_mappings (catalog_prefix, spec, detail) values
+                ($2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3),
```
It feels weird to hard-code the storage bucket for this, especially given the fact that the accounts email is parameterized. Long term, I'd think that the storage bucket would be determined based on which data-plane the tenant will be mapped to. I don't know if it's reasonable to go straight to looking up the bucket based on the data-plane mapping, but it seems at least worth considering. If you did, it might also allow eliminate the storage mapping concerns that motivate doing this as an offline migration. Otherwise, perhaps just making it an application argument would be the way to go.

jgraettinger replied:
:shrug: I'd like to KISS for now. Yes, this is a clear target for refactor when there's more built out to support creating storage mappings, etc.

psFried reviewed (changes requested):
I don't see any problems with the migration, but I do want to just double check the plan for applying it. Deleting the storage mappings will cause all builds to fail until the agent adds them in again. But at _that_ point, I'd think that the builds would succeed, but use the wrong storage mapping, until the `ops/` storage mapping is set back to its original value. Does that sound right? If so, then it sounds like we should ensure that nobody is using the system during this process.

jshearer commented on supabase/pending/directives.sh:
Should this be automatic?

jgraettinger replied:
No, this is a note to myself when actually doing the deploy. I don't think it's worth further time investment.

jshearer commented on supabase/pending/directives.sh:
```diff
@@ -0,0 +1,96 @@
+drop table public.task_stats_by_minute cascade;
+drop table public.flow_checkpoints_v1 cascade;
+drop table public.task_stats_by_hour cascade;
+drop table public.task_stats_by_day cascade;
```
Don't these table not exist on prod yet? So don't we need `drop table if exists` here, otherwise this will fail?

jgraettinger replied:
They do exist in prod, and we don't want them to.

jgraettinger commented on ops-catalog/template-common.flow.yaml:
```diff
@@ -0,0 +1,111 @@
+            openSecondsTotal: 0.03
+
+storageMappings:
+  "": { stores: [{ provider: S3, bucket: a-bucket }] }
```
only for local testing of the `ops-catalog` itself. We don't push these up to the control plane API. Commented.

jgraettinger commented on supabase/pending/directives.sh:
```diff
@@ -0,0 +1,96 @@
+    cat <<EOF
+begin;
+
+drop table public.flow_materializations_v2 cascade;
```
this is a table that's created by our `materialize-postgres` connector as part of its function, and isn't part of our intended / desired control-plane API.

It's desirable cleanup to get rid of it (it's moving into the `task_stat_partitions` schema).

jgraettinger commented:
Thanks all, I'm going to go ahead, merge, and rip this particular band-aide off.
